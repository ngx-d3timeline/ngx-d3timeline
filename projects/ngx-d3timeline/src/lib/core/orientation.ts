export enum Orientation {
  Horizontal = 'Horizontal',
  Vertical = 'Vertical'
}

export function flipOrientation(orientation: Orientation): Orientation {
  return orientation === Orientation.Vertical
    ? Orientation.Horizontal
    : Orientation.Vertical;
}
