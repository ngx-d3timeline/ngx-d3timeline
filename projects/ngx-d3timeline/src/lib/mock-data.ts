import { TimelineEvent } from './timeline-event';

export const mockData: TimelineEvent[] = [
  {
    id: 1,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 2,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 3,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 1',
    sequence: 1
  },
  {
    id: 13,
    start: new Date('2017-09-22T10:00:00'),
    finish: new Date('2017-09-22T10:08:31'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 14,
    start: new Date('2017-09-22T09:15:00'),
    finish: new Date('2017-09-22T09:16:16'),
    type: 'Waiting',
    series: 'Series 2',
    sequence: 6
  },
  {
    id: 15,
    start: new Date('2017-09-22T13:45:00'),
    finish: new Date('2017-09-22T14:52:27'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 41,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 42,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 43,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 3',
    sequence: 1
  }
];
