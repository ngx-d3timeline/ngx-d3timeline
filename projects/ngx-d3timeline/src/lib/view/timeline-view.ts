export class TimelineView {
  readonly width: number;
  readonly height: number;
  readonly margin: number;

  readonly bottom: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;

  constructor({
    width,
    height,
    margin
  }: {
    width: number;
    height: number;
    margin: number;
  }) {
    this.width = width;
    this.height = height;
    this.margin = margin;

    this.bottom = height - margin;
    this.left = 0;
    this.right = width - margin;
    this.top = 0;
  }
}
