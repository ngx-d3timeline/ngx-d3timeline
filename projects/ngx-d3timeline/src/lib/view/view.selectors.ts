import { selectView } from '../store/state';
import {
  getViewTopLeft,
  getViewClipRectangle,
  getViewClipRectangleHeight,
  getViewClipRectangleWidth
} from './view-utils';
import { createSelector } from '../store-lib/selector/create-selector';

export const selectViewTopLeft = createSelector(selectView, getViewTopLeft);

export const selectViewClipRectangleHeight = createSelector(
  selectView,
  getViewClipRectangleHeight
);

export const selectViewClipRectangleWidth = createSelector(
  selectView,
  getViewClipRectangleWidth
);

export const selectViewClipRectangle = createSelector(
  selectViewTopLeft,
  selectViewClipRectangleWidth,
  selectViewClipRectangleHeight,
  getViewClipRectangle
);
