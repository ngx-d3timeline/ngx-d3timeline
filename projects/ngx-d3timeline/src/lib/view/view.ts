export class View {
  readonly width: number;
  readonly height: number;
  readonly margin = 50;

  readonly left = this.margin;
  readonly top = this.margin;
  readonly right: number;
  readonly bottom: number;

  constructor([width, height]: [number, number]) {
    this.width = width;
    this.height = height;

    this.bottom = height - this.margin;
    this.right = width - this.margin;
  }
}
