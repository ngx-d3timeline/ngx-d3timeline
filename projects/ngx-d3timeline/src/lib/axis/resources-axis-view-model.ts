import { ScaleBand, scaleBand } from 'd3-scale';
import { TimelineEvent } from '../timeline-event';
import { TimelineView } from '../view/timeline-view';

export class ResourcesAxisViewModel {
  private readonly scaleBand: ScaleBand<string>;

  constructor(data: TimelineEvent[], timelineView: TimelineView) {
    this.scaleBand = this.configureScaleBand(data, timelineView);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleBand.domain().map(tick => ({
      label: tick,
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleBand.range()[1];
  }

  get bandwidth() {
    return this.scaleBand.bandwidth();
  }

  private configureScaleBand(
    data: TimelineEvent[],
    timelineView: TimelineView
  ) {
    return scaleBand()
      .domain([...new Set(data.map(d => d.series))])
      .range([timelineView.bounds.left, timelineView.bounds.right]);
  }

  private tickTransform(tick: string) {
    return `translate(${this.getBandMidPoint(tick)}, 0)`;
  }

  private getBandMidPoint(tick: string) {
    return this.scaleBand(tick) + this.scaleBand.bandwidth() / 2;
  }
}
