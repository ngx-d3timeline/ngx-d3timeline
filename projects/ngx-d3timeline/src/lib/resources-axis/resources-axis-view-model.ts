import { ScaleBand, scaleBand } from 'd3-scale';
import { TimelineEvent } from '../timeline-event';
import { TimelineView } from '../view/timeline-view';
import { Orientation } from '../orientation';

export class ResourcesAxisViewModel {
  private readonly scaleBand: ScaleBand<string>;

  constructor(
    data: TimelineEvent[],
    timelineView: TimelineView,
    private readonly orientation: Orientation
  ) {
    this.scaleBand = this.configureScaleBand(data, timelineView);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleBand.domain().map(tick => ({
      label: tick,
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleBand.range()[1];
  }

  get bandwidth() {
    return this.scaleBand.bandwidth();
  }

  private configureScaleBand(
    data: TimelineEvent[],
    timelineView: TimelineView
  ) {
    return scaleBand()
      .domain([...new Set(data.map(d => d.series))])
      .range(this.getRange(timelineView.bounds));
  }

  private getRange(bounds: any): [number, number] {
    return this.orientation === Orientation.Vertical
      ? [bounds.left, bounds.right]
      : [bounds.top, bounds.bottom];
  }

  private tickTransform(tick: string) {
    return this.orientation === Orientation.Vertical
      ? `translate(${this.getBandMidPoint(tick)}, 0)`
      : `translate(0, ${this.getBandMidPoint(tick)})`;
  }

  private getBandMidPoint(tick: string) {
    return this.mapToPixel(tick) + this.scaleBand.bandwidth() / 2;
  }

  mapToPixel(domain: string): number {
    return this.scaleBand(domain);
  }
}
