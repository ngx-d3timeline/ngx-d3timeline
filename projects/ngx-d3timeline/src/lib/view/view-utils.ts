import { View } from './view';
import { Point } from '../core/point';
import { Rectangle } from '../core/rectangle';

export function getViewTopLeft(view: View): Point {
  return { x: view.left, y: view.top };
}

export function getViewClipRectangle(view: View): Rectangle {
  return {
    width: view.width - 2 * view.margin,
    height: view.height - 2 * view.margin,
    x: view.left,
    y: view.top
  };
}
