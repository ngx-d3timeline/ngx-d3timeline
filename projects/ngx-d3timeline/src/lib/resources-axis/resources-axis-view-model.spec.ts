import { ResourcesAxisViewModel } from './resources-axis-view-model';
import { mockData } from '../mock-data';
import { TimelineView } from '../view/timeline-view';
import { Orientation } from '../orientation';

describe('ResourcesAxisViewModel', () => {
  let vm: ResourcesAxisViewModel;
  const timelineView: TimelineView = new TimelineView(300, 400, 50);

  describe('when Vertical', () => {
    beforeEach(() => {
      vm = new ResourcesAxisViewModel(
        mockData,
        timelineView,
        Orientation.Vertical
      );
    });

    it('should return correct tickInfo', () => {
      expect(vm.tickInfos).toMatchSnapshot();
    });

    it('should return correct rangeLimit', () => {
      expect(vm.rangeLimit).toBe(timelineView.bounds.right);
    });
  });

  describe('when Horizontal', () => {
    beforeEach(() => {
      vm = new ResourcesAxisViewModel(
        mockData,
        timelineView,
        Orientation.Horizontal
      );
    });

    it('should return correct tickInfo', () => {
      expect(vm.tickInfos).toMatchSnapshot();
    });

    it('should return correct rangeLimit', () => {
      expect(vm.rangeLimit).toBe(timelineView.bounds.bottom);
    });
  });
});
