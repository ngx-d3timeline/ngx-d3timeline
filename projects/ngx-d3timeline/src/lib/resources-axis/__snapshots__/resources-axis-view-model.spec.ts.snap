// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourcesAxisViewModel when Horizontal should return correct tickInfo 1`] = `
Array [
  Object {
    "label": "Series 1",
    "transform": "translate(0, 58.333333333333336)",
  },
  Object {
    "label": "Series 2",
    "transform": "translate(0, 175)",
  },
  Object {
    "label": "Series 3",
    "transform": "translate(0, 291.6666666666667)",
  },
]
`;

exports[`ResourcesAxisViewModel when Vertical should return correct tickInfo 1`] = `
Array [
  Object {
    "label": "Series 1",
    "transform": "translate(41.666666666666664, 0)",
  },
  Object {
    "label": "Series 2",
    "transform": "translate(125, 0)",
  },
  Object {
    "label": "Series 3",
    "transform": "translate(208.33333333333331, 0)",
  },
]
`;
